// ipreview/example/src/pages/index.js
import React from 'react'
import { ImgPreview } from '../../../src/index.js'


export default class extends React.Component {

    constructor(props) {
        super(props)
        this.oridinary = React.createRef()
        this.download = React.createRef()
        this.many = React.createRef()
        this.concise = React.createRef()
        this.custom = React.createRef()
        this.state = {
            iconPosition: 'left',
            modes: [
                {
                    mode: '普通模式',
                    ref: this.oridinary,
                },
                {
                    mode: '可供下载',
                    ref: this.download,
                },
                {
                    mode: '多图模式',
                    ref: this.many,
                },
                {
                    mode: '简洁模式',
                    ref: this.concise,
                },
                {
                    mode: '自定义图标',
                    ref: this.custom,
                }, 
            ],
            activeMode: '普通模式',
            iPreviewApi: [
                {
                    arg: 'concise',
                    type: 'boolean',
                    default: 'false',
                    intro: '是否开启简洁模式',
                },
                {
                    arg: 'download',
                    type: 'boolean',
                    default: 'false',
                    intro: '是否提供下载选项',
                },
                {
                    arg: 'url',
                    type: 'string',
                    default: '无',
                    intro: '必填，预览的图片的url',
                },
                {
                    arg: 'urlList',
                    type: 'array',
                    default: '[]',
                    intro: '选填，多图模式下的图片的url组成的数组，填写此字段的同时url字段不可为空，可以使用\'\'进行占位，若同时填写了此字段与url字段，则url字段不起作用',
                },
                {
                    arg: 'zIndex',
                    type: 'number',
                    default: '999',
                    intro: '预览图层的z-index值，默认是999，可根据实际场景进行调整',
                },
                {
                    arg: 'iconSet',
                    type: 'object',
                    default: `{
                        close: null,
                        pre: null,
                        next: null,
                        rotateLeft: null,
                        rotateRight: null,
                        large: null,
                        small: null,
                        download: null,
                        reset: null
                    }`,
                    intro: '预览图层中的icon集合，该对象可只填部分属性，属性取值为react节点',
                }
            ]
        }
    }

    changeMode(res) {
        let { modes } = this.state
        this.setState({
            activeMode: res
        })
        window.scrollTo(0, modes.find(item => item.mode === res).ref.current.offsetTop)
    }

    render() {
        const { modes, activeMode, iPreviewApi } = this.state
        let url = 'https://drscdn.500px.org/photo/63978593/q%3D80_h%3D600/v2?sig=bfa4ebd5632d31fea893a35facc0a920932b1ebe4b2fe7a5ff3a948f9f2a69e2'
        let urlList = [
            'https://drscdn.500px.org/photo/63978593/q%3D80_h%3D600/v2?sig=bfa4ebd5632d31fea893a35facc0a920932b1ebe4b2fe7a5ff3a948f9f2a69e2',
            'https://drscdn.500px.org/photo/65283001/q%3D80_h%3D600/v2?sig=fb397ea1715c528db2569228f14774aa79c6b98c1e45e4d66abdc60db23a59d2',
            'https://drscdn.500px.org/photo/64682967/q%3D80_h%3D600/v2?sig=5091e0cca8c401b745b971221694eb991c8f2d64facf62afdb37c32dd91d46c3',
            'https://drscdn.500px.org/photo/125902791/q%3D80_h%3D600/v2?sig=a244dc70da63d77b4bc3d658322dd16891b184089a83492aa3ca2ba7d25812e4'
        ]
        const iconSet = {
            close: <svg t="1575425721236" className="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="24960" width="200" height="200"><path d="M509.866667 32C245.333333 32 32 247.466667 32 512s213.333333 480 477.866667 480S987.733333 776.533333 987.733333 512 774.4 32 509.866667 32z m0 896C281.6 928 96 742.4 96 512S281.6 96 509.866667 96 923.733333 281.6 923.733333 512s-185.6 416-413.866666 416z" fill="#ffffff" p-id="24961"></path><path d="M693.333333 330.666667c-12.8-12.8-32-12.8-44.8 0L512 467.2l-136.533333-136.533333c-12.8-12.8-32-12.8-44.8 0-12.8 12.8-12.8 32 0 44.8l136.533333 136.533333-136.533333 136.533333c-12.8 12.8-12.8 32 0 44.8 6.4 6.4 14.933333 8.533333 23.466666 8.533334s17.066667-2.133333 23.466667-8.533334l136.533333-136.533333 136.533334 136.533333c6.4 6.4 14.933333 8.533333 23.466666 8.533334s17.066667-2.133333 23.466667-8.533334c12.8-12.8 12.8-32 0-44.8L556.8 512l136.533333-136.533333c12.8-12.8 12.8-32 0-44.8z" fill="#ffffff" p-id="24962"></path></svg>,
            pre: <svg t="1575425862709" className="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="27209" width="200" height="200"><path d="M759.931232 12.714422L265.535817 419.575931v-342.311366A41.077364 41.077364 0 1 0 182.892073 76.286533v870.937918a41.077364 41.077364 0 1 0 82.643744 0v-345.245463l494.395415 409.30659a48.901624 48.901624 0 0 0 81.176695-36.187202V48.901624A48.901624 48.901624 0 0 0 759.931232 12.714422z" fill="#ffffff" p-id="27210"></path></svg>,
            next: <svg t="1575425838097" className="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="26748" width="200" height="200"><path d="M263.889466 1011.570793l494.343464-407.339014v343.498087a41.242369 41.242369 0 1 0 82.484738 0V76.555441a41.242369 41.242369 0 1 0-82.484738 0v345.757943L264.45443 12.714514A48.5869 48.5869 0 0 0 183.09962 48.872207v926.540893a48.5869 48.5869 0 0 0 80.789846 36.157693z" fill="#ffffff" p-id="26749"></path></svg>,
            rotateLeft: <svg t="1575426047671" className="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="29476" width="200" height="200"><path d="M924.8 337.6c-22.6-53.4-54.9-101.3-96-142.4s-89-73.4-142.4-96C631.1 75.9 572.5 64 512 64S392.9 75.9 337.6 99.2c-53.4 22.6-101.3 54.9-142.4 96-22.4 22.4-42.2 46.8-59.2 73.1V228c0-19.8-16.2-36-36-36s-36 16.2-36 36v288c0 19.8 16.2 36 36 36s36-16.2 36-36v-50.2c4.2-34.8 13.2-68.7 27-101.2 19.1-45.1 46.4-85.6 81.2-120.4C279 209.4 319.5 182 364.6 163c46.7-19.7 96.3-29.8 147.4-29.8 51.2 0 100.8 10 147.4 29.8 45.1 19.1 85.6 46.4 120.4 81.2C814.6 279 842 319.5 861 364.6c19.7 46.7 29.8 96.3 29.8 147.4 0 51.2-10 100.8-29.8 147.4-19.1 45.1-46.4 85.6-81.2 120.4C745 814.6 704.5 842 659.4 861c-46.7 19.7-96.3 29.8-147.4 29.8-64.6 0-128.4-16.5-184.4-47.8-54.4-30.4-100.9-74.1-134.6-126.6-10.3-16.1-31.7-20.8-47.8-10.4-16.1 10.3-20.8 31.7-10.4 47.8 39.8 62 94.8 113.7 159.1 149.6 66.2 37 141.7 56.6 218.1 56.6 60.5 0 119.1-11.9 174.4-35.2 53.4-22.6 101.3-54.9 142.4-96 41.1-41.1 73.4-89 96-142.4C948.1 631.1 960 572.5 960 512s-11.9-119.1-35.2-174.4z" p-id="29477" fill="#ffffff"></path><path d="M275.4 575.5c9.5-2.5 19.1 2.9 22.3 12.2 3.5 10.2 9.9 17.7 19.1 22.6 7.1 3.9 15.1 5.8 24 5.8 16.6 0 30.8-6.9 42.5-20.8 11.7-13.8 20-32.7 24.9-75.1-7.7 12.2-17.3 20.8-28.7 25.8-11.4 5-23.7 7.4-36.8 7.4-26.7 0-47.7-8.3-63.3-24.9-15.5-16.6-23.3-37.9-23.3-64.1 0-25.1 7.7-47.1 23-66.2 15.3-19 37.9-28.6 67.8-28.6 40.3 0 68.1 18.1 83.4 54.4 8.5 19.9 12.7 44.9 12.7 74.9 0 33.8-5.1 63.8-15.3 89.9-16.9 43.5-45.5 65.2-85.8 65.2-27 0-47.6-7.1-61.6-21.2-10-10.1-16.4-22-19.3-35.8-2-9.6 4-19.1 13.5-21.6l0.9 0.1z m103-74.4c9.4-7.5 14.1-20.6 14.1-39.3 0-16.8-4.2-29.3-12.7-37.5S360.6 412 347.5 412c-14 0-25.2 4.7-33.4 14.1-8.2 9.4-12.4 22-12.4 37.7 0 14.9 3.6 26.7 10.9 35.5 7.2 8.8 18.8 13.1 34.6 13.1 11.4 0 21.8-3.8 31.2-11.3zM646.6 414.4c12.4 22.8 18.5 54 18.5 93.7 0 37.6-5.6 68.7-16.8 93.3-16.2 35.3-42.8 52.9-79.6 52.9-33.2 0-57.9-14.4-74.2-43.3-13.5-24.1-20.3-56.4-20.3-97 0-31.4 4.1-58.4 12.2-80.9 15.2-42 42.7-63 82.5-63 35.9 0 61.8 14.8 77.7 44.3z m-40.2 173.3c9.4-13.9 14-39.9 14-78 0-27.4-3.4-50-10.1-67.7-6.8-17.7-19.9-26.6-39.4-26.6-17.9 0-31 8.4-39.3 25.2-8.3 16.8-12.4 41.6-12.4 74.3 0 24.6 2.6 44.4 7.9 59.4 8.1 22.8 22 34.3 41.6 34.3 15.7 0 28.3-7 37.7-20.9zM803.3 387.2c11.2 11.3 16.8 25 16.8 41.2 0 16.7-5.8 30.7-17.5 41.8C791 481.4 777.4 487 762 487c-17.1 0-31.2-5.8-42.1-17.4-10.9-11.6-16.4-25.1-16.4-40.6 0-16.5 5.8-30.4 17.3-41.7 11.5-11.3 25.3-17 41.2-17 16.3 0 30.1 5.7 41.3 16.9zM739.5 451c6.2 6.2 13.7 9.3 22.5 9.3 8.4 0 15.8-3.1 22.1-9.3 6.3-6.2 9.4-13.7 9.4-22.6 0-8.5-3.1-15.9-9.3-22.1-6.2-6.2-13.6-9.3-22.2-9.3s-16.1 3.1-22.4 9.3c-6.3 6.2-9.4 13.7-9.4 22.6-0.1 8.4 3 15.8 9.3 22.1z" p-id="29478" fill="#ffffff"></path></svg>,
            rotateRight: <svg t="1575426077079" className="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="29634" width="200" height="200"><path d="M1464.3 279.7" p-id="29635" fill="#ffffff"></path><path d="M512 960c-60.5 0-119.1-11.9-174.4-35.2-53.4-22.6-101.3-54.9-142.4-96s-73.4-89-96-142.4C75.9 631.1 64 572.5 64 512s11.9-119.1 35.2-174.4c22.6-53.4 54.9-101.3 96-142.4s89-73.4 142.4-96C392.9 75.9 451.5 64 512 64s119.1 11.9 174.4 35.2c53.4 22.6 101.3 54.9 142.4 96s73.4 89 96 142.4C948.1 392.9 960 451.5 960 512c0 19.1-15.5 34.6-34.6 34.6s-34.6-15.5-34.6-34.6c0-51.2-10-100.8-29.8-147.4-19.1-45.1-46.4-85.6-81.2-120.4C745 209.4 704.5 182 659.4 163c-46.7-19.7-96.3-29.8-147.4-29.8-51.2 0-100.8 10-147.4 29.8-45.1 19.1-85.6 46.4-120.4 81.2S182 319.5 163 364.6c-19.7 46.7-29.8 96.3-29.8 147.4 0 51.2 10 100.8 29.8 147.4 19.1 45.1 46.4 85.6 81.2 120.4C279 814.6 319.5 842 364.6 861c46.7 19.7 96.3 29.8 147.4 29.8 64.6 0 128.4-16.5 184.4-47.8 54.4-30.4 100.9-74.1 134.6-126.6 10.3-16.1 31.7-20.8 47.8-10.4 16.1 10.3 20.8 31.7 10.4 47.8-39.8 62-94.8 113.7-159.1 149.6-66.2 37-141.7 56.6-218.1 56.6z" p-id="29636" fill="#ffffff"></path><path d="M924 552c-19.8 0-36-16.2-36-36V228c0-19.8 16.2-36 36-36s36 16.2 36 36v288c0 19.8-16.2 36-36 36zM275.4 575.5c9.5-2.5 19.1 2.9 22.3 12.2 3.5 10.2 9.9 17.7 19.1 22.6 7.1 3.9 15.1 5.8 24 5.8 16.6 0 30.8-6.9 42.5-20.8 11.7-13.8 20-32.7 24.9-75.1-7.7 12.2-17.3 20.8-28.7 25.8-11.4 5-23.7 7.4-36.8 7.4-26.7 0-47.7-8.3-63.3-24.9-15.5-16.6-23.3-37.9-23.3-64.1 0-25.1 7.7-47.1 23-66.2 15.3-19 37.9-28.6 67.8-28.6 40.3 0 68.1 18.1 83.4 54.4 8.5 19.9 12.7 44.9 12.7 74.9 0 33.8-5.1 63.8-15.3 89.9-16.9 43.5-45.5 65.2-85.8 65.2-27 0-47.6-7.1-61.6-21.2-10-10.1-16.4-22-19.3-35.8-2-9.6 4-19.1 13.5-21.6l0.9 0.1z m103-74.4c9.4-7.5 14.1-20.6 14.1-39.3 0-16.8-4.2-29.3-12.7-37.5S360.6 412 347.5 412c-14 0-25.2 4.7-33.4 14.1-8.2 9.4-12.4 22-12.4 37.7 0 14.9 3.6 26.7 10.9 35.5 7.2 8.8 18.8 13.1 34.6 13.1 11.4 0 21.8-3.8 31.2-11.3zM646.6 414.4c12.4 22.8 18.5 54 18.5 93.7 0 37.6-5.6 68.7-16.8 93.3-16.2 35.3-42.8 52.9-79.6 52.9-33.2 0-57.9-14.4-74.2-43.3-13.5-24.1-20.3-56.4-20.3-97 0-31.4 4.1-58.4 12.2-80.9 15.2-42 42.7-63 82.5-63 35.9 0 61.8 14.8 77.7 44.3z m-40.2 173.3c9.4-13.9 14-39.9 14-78 0-27.4-3.4-50-10.1-67.7-6.8-17.7-19.9-26.6-39.4-26.6-17.9 0-31 8.4-39.3 25.2-8.3 16.8-12.4 41.6-12.4 74.3 0 24.6 2.6 44.4 7.9 59.4 8.1 22.8 22 34.3 41.6 34.3 15.7 0 28.3-7 37.7-20.9zM803.3 387.2c11.2 11.3 16.8 25 16.8 41.2 0 16.7-5.8 30.7-17.5 41.8C791 481.4 777.4 487 762 487c-17.1 0-31.2-5.8-42.1-17.4-10.9-11.6-16.4-25.1-16.4-40.6 0-16.5 5.8-30.4 17.3-41.7 11.5-11.3 25.3-17 41.2-17 16.3 0 30.1 5.7 41.3 16.9zM739.5 451c6.2 6.2 13.7 9.3 22.5 9.3 8.4 0 15.8-3.1 22.1-9.3 6.3-6.2 9.4-13.7 9.4-22.6 0-8.5-3.1-15.9-9.3-22.1-6.2-6.2-13.6-9.3-22.2-9.3s-16.1 3.1-22.4 9.3c-6.3 6.2-9.4 13.7-9.4 22.6-0.1 8.4 3 15.8 9.3 22.1z" p-id="29637" fill="#ffffff"></path></svg>,
            large: <svg t="1575426117929" className="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="30398" width="200" height="200"><path d="M352 896a32 32 0 1 1 0 64h-256a32 32 0 0 1-32-32v-256a32 32 0 1 1 64 0v205.12l259.776-259.776a32 32 0 1 1 45.248 45.312L199.616 896H352z m320-768a32 32 0 1 1 0-64h256a32 32 0 0 1 32 32v256a32 32 0 1 1-64 0V173.248l-227.776 227.776a32 32 0 0 1-45.248-45.248L850.752 128H672z" p-id="30399" fill="#ffffff"></path></svg>,
            reset: <svg t="1575455629635" className="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="32486" width="200" height="200"><path d="M511.964163 0.000213A515.803894 515.803894 0 0 0 85.32736 229.104176v-106.6592a42.66368 42.66368 0 1 0-85.32736 0v213.318401a42.66368 42.66368 0 0 0 42.66368 42.66368h213.318401a42.66368 42.66368 0 0 0 0-85.327361H145.269831a426.636802 426.636802 0 1 1-9.599328 419.383977 42.66368 42.66368 0 0 0-75.088077 40.103859A511.964163 511.964163 0 1 0 511.964163 0.000213z" fill="#ffffff" p-id="32487"></path></svg>,
            small: <svg t="1575426167719" className="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="31735" width="200" height="200"><path d="M512 768.141c0 17.51-14.508 32.018-32.018 32.018-8.505 0-16.509-3.502-22.512-9.505l-72.04-72.04L219.34 884.705c-3.002 3.001-7.505 5.002-11.507 5.002s-8.505-2.001-11.506-5.002l-57.031-57.032c-3.002-3.001-5.003-7.504-5.003-11.506s2.001-8.505 5.003-11.506l166.091-166.092-72.039-72.039c-6.004-6.004-9.506-14.008-9.506-22.513 0-17.51 14.508-32.018 32.018-32.018l224.123 0c17.51 0 32.018 14.508 32.018 32.018L512.001 768.141zM884.705 219.34 718.613 385.431l72.04 72.04c6.003 6.003 9.505 14.007 9.505 22.512 0 17.51-14.508 32.018-32.018 32.018L544.018 512.001C526.508 512 512 497.492 512 479.982L512 255.859c0-17.51 14.508-32.018 32.018-32.018 8.505 0 16.51 3.502 22.513 9.506l72.039 72.039L804.66 139.296c3.002-3.002 7.505-5.003 11.507-5.003s8.505 2.001 11.507 5.003l57.031 57.031c3.001 3.001 5.002 7.504 5.002 11.506S887.706 216.338 884.705 219.34z" p-id="31736" fill="#ffffff"></path></svg>
        }
        return (
            <div className="demo-wrapper">
                <div className="example-header">Ipreview/图片预览</div>
                <div className="example-nav">
                    {modes.map(item => (
                        <div className={`nav-item${item.mode === activeMode ? ' nav-item-active' : ''}`} key={item.mode} onClick={() => {this.changeMode(item.mode)}}>{item.mode}</div>
                    ))}
                </div>
                <div className="example-wrapper">
                    <p ref={this.oridinary} className="mode-title">普通模式</p>
                    <ImgPreview url={url}><img src={url}/></ImgPreview>
                    <p ref={this.download} className="mode-title">可供下载</p>
                    <ImgPreview download={true} url={url}><img src={url}/></ImgPreview>
                    <p ref={this.many} className="mode-title">多图模式</p>
                    <ImgPreview download={true} url={''} urlList={urlList}><img src={url}/></ImgPreview>
                    <p ref={this.concise} className="mode-title">简洁模式</p>
                    <ImgPreview concise={true} url={url}><img src={url}/></ImgPreview>
                    <p ref={this.custom} className="mode-title">自定义图标</p>
                    <ImgPreview  iconSet={iconSet} url={url} urlList={urlList}><img src={url}/></ImgPreview>
                    <p className="mode-title">API</p>
                    <table>
                        <thead>
                            <tr>
                                <th>ImgPreview</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr>
                                <td>参数</td>
                                <td>类型</td>
                                <td>默认值</td>
                                <td>说明</td>
                            </tr>
                            {iPreviewApi.map(item => <tr key={item.arg}>
                            <td>{item.arg}</td>
                            <td>{item.type}</td>
                            <td>{item.default}</td>
                            <td>{item.intro}</td>
                            </tr>)}
                        </tbody>
                    </table>
                </div>
            </div>
        )
    }
}
